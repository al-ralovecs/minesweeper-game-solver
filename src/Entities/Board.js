export default class Board {
  constructor(state) {
    this.state = state;

    this.tryFlagMines();
  }

  get height() {
    return this.state.length;
  }

  get width() {
    if (typeof this.boardWidth === 'undefined') {
      this.boardWidth = this.getBoardWidth();
    }

    return this.boardWidth;
  }

  getBoardWidth() {
    const width = this.state[0].length;

    if (this.height === 1) {
      return width;
    }

    for (let i = 1; i < this.height; i++) {
      if (width !== this.state[i].length) {
        throw Error();
      }
    }

    return width;
  }

  tryFlagMines() {
    let flags = [];

    for (let i = 0; i < this.height; i++) {
      let row = [];
      for (let j = 0; j < this.width; j++) {
        row.push(false);
      }
      flags.push(row);
    }

    for (let i = 0; i < this.height; i++) {
      for (let j = 0; j < this.width; j++) {
        let curNum = this.getCell(i, j);

        if (curNum < 1 || curNum !== this.countFreeSquaresAround(i, j)) {
          continue;
        }

        for (let ii = 0; ii < this.height; ii++) {
          for (let jj = 0; jj < this.width; jj++) {
            if (Math.abs(ii-i) <= 1 && Math.abs(jj-j) <= 1) {
              if (this.getCell(ii, jj) === -1 && ! flags[ii][jj]) {
                flags[ii][jj] = true;
              }
            }
          }
        }
      }
    }

    this.flags = flags;
  }

  getCell(y, x) {
    if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
      return this.state[y][x];
    }

    return -10;
  }

  hasMine(y, x) {
    if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
      return this.flags[y][x];
    }

    throw Error();
  }

  isBoundary(y, x) {
    if (this.getCell(y, x) !== -1) {
      return false;
    }

    const
      oU = y === 0,
      oD = y === (this.height - 1),
      oL = x === 0,
      oR = x === (this.width - 1);

    if (!oU && this.getCell(y-1, x) >= 0) {
      return true;
    }
    if (!oL && this.getCell(y, x-1) >= 0) {
      return true;
    }
    if (!oD && this.getCell(y+1, x) >= 0) {
      return true;
    }
    if (!oR && this.getCell(y, x+1) >= 0) {
      return true;
    }
    if (!oU && !oL && this.getCell(y-1, x-1) >= 0) {
      return true;
    }
    if (!oU && !oR && this.getCell(y-1, x+1) >= 0) {
      return true;
    }
    if (!oD && !oL && this.getCell(y+1, x-1) >= 0) {
      return true;
    }
    if (!oD && !oR && this.getCell(y+1, x+1) >= 0) {
      return true;
    }

    return false;
  }

  countFreeSquaresAround(y, x) {
    let result = 0;

    if (this.getCell(y-1, x) === -1) {
      result++;
    }
    if (this.getCell(y+1, x) === -1) {
      result++;
    }
    if (this.getCell(y, x-1) === -1) {
      result++;
    }
    if (this.getCell(y, x+1) === -1) {
      result++;
    }
    if (this.getCell(y-1, x-1) === -1) {
      result++;
    }
    if (this.getCell(y-1, x+1) === -1) {
      result++;
    }
    if (this.getCell(y+1, x-1) === -1) {
      result++;
    }
    if (this.getCell(y+1, x+1) === -1) {
      result++;
    }

    return result;
  }
}
