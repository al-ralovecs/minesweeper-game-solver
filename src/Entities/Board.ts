export default class Board {
    private readonly state: number[][];
    private flags: boolean[][];

    private height: number;
    private width: number;

    public constructor(state: number[][])
    {
        if (! Board.validate(state)) {
            throw Error();
        }

        this.state = state;

        this.flagMines();
    }

    public get getHeight(): number
    {
        if (typeof this.height === 'undefined') {
            this.height = this.state.length;
        }

        return this.height;
    }

    public get getWidth(): number
    {
        if (typeof this.width === 'undefined') {
            this.width = this.state[0].length;
        }

        return this.width;
    }

    public getValue(y: number, x: number): number
    {
        if ( y >= 0 && y < this.getHeight && x >= 0 && x < this.getWidth) {
            return this.state[y][x];
        }

        return -10;
    }

    public getSurroundingFreeSquaresCount(y: number, x: number): number
    {
        let result = 0;

        if (this.getValue(y-1, x) === -1) {
            result++;
        }
        if (this.getValue(y+1, x) === -1) {
            result++;
        }
        if (this.getValue(y, x-1) === -1) {
            result++;
        }
        if (this.getValue(y, x+1) === -1) {
            result++;
        }
        if (this.getValue(y-1, x-1) === -1) {
            result++;
        }
        if (this.getValue(y-1, x+1) === -1) {
            result++;
        }
        if (this.getValue(y+1, x-1) === -1) {
            result++;
        }
        if (this.getValue(y+1, x+1) === -1) {
            result++;
        }

        return result;
    }

    public isBoundary(y: number, x: number): boolean
    {
        if (this.getValue(y, x) !== -1) {
            return false;
        }

        const
            oU: boolean = y === 0,
            oD: boolean = y === (this.getHeight - 1),
            oL: boolean = x === 0,
            oR: boolean = x === (this.getWidth - 1);

        if (!oU && this.getValue(y - 1, x) >= 0) {
            return true;
        }
        if (!oL && this.getValue(y, x - 1) >= 0) {
            return true;
        }
        if (!oD && this.getValue(y + 1, x) >= 0) {
            return true;
        }
        if (!oR && this.getValue(y, x + 1) >= 0) {
            return true;
        }
        if (!oU && !oL && this.getValue(y - 1, x - 1) >= 0) {
            return true;
        }
        if (!oU && !oR && this.getValue(y - 1, x + 1) >= 0) {
            return true;
        }
        if (!oD && !oL && this.getValue(y + 1, x - 1) >= 0) {
            return true;
        }

        return !oD && !oR && this.getValue(y + 1, x + 1) >= 0;
    }

    public hasMine(y: number, x: number): boolean
    {
        if (y >= 0 && y < this.getHeight && x >= 0 && x < this.getWidth) {
            return this.flags[y][x];
        }

        throw Error();
    }

    private flagMines(): void
    {
        let flags: boolean[][] = [];

        for (let i = 0; i < this.getHeight; i++) {
            let row: boolean[] = [];
            for (let j = 0; j < this.getWidth; j++) {
                row.push(false);
            }
            flags.push(row);
        }

        for (let i: number = 0; i < this.height; i++) {
            for (let j: number = 0; j < this.width; j++) {
                let currentValue: number = this.getValue(i, j);

                if (currentValue < 1 || currentValue !== this.getSurroundingFreeSquaresCount(i, j)) {
                    continue;
                }

                for (let ii: number = 0; ii < this.getHeight; ii++) {
                    for (let jj:number = 0; jj < this.getWidth; jj++) {
                        if (Math.abs(ii-i) <= 1 && Math.abs(jj-j) <= 1) {
                            if (this.getValue(ii, jj) === -1 && ! flags[ii][jj]) {
                                flags[ii][jj] = true;
                            }
                        }
                    }
                }
            }
        }

        this.flags = flags;
    }

    private static validate(state: number[][]): boolean
    {
        let width: number = 0;
        for (let i = 0; i < state.length; i++) {
            if (0 === width) {
                width = state[i].length;
                continue;
            }

            if (state[i].length === width) {
                continue;
            }

            return false;
        }

        return true;
    }
}
