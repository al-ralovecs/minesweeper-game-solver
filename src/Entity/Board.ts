import BitwiseMatrix from './BitwiseMatrix';
import countUnrevealedItemsAround from '../Routine/board.item.countUnrevealedItemsAround';
import flagEmpty from '../Routine/board.matrix.flagEmpty';
import flagMines from '../Routine/board.matrix.flagMines';
import getIsBoundary from '../Routine/board.item.getIsBoundary';
import getTotalItemsAround from '../Routine/board.item.getTotalItemsAround';
import getValue from '../Routine/board.item.getValue';
import isValid from '../Routine/board.matrix.isValid';

export default class Board {
    private readonly data: number[][];

    public readonly height: number;
    public readonly width: number;

    private modelMines: BitwiseMatrix;
    private modelEmpty: BitwiseMatrix;

    public constructor(data: number[][], init: boolean = true)
    {
        if (! Board.isValid(data)) {
            throw Error();
        }

        this.data = data;
        this.height = data.length;
        this.width = data[0].length;

        if (init) {
            this.modelMines = new BitwiseMatrix(this.height, this.width);
            this.modelEmpty = new BitwiseMatrix(this.height, this.width);

            this.modelMines.setData = flagMines(this.modelMines.getData, this.data, this.height, this.width);
            this.modelEmpty.setData = flagEmpty(this.modelEmpty.getData, this.data, this.height, this.width);
        }
    }

    public get getMinesMap(): BitwiseMatrix
    {
        return this.modelMines;
    }

    public set setMinesMap(modelMines: BitwiseMatrix)
    {
        this.modelMines = modelMines;
    }

    public get getEmptyMap(): BitwiseMatrix
    {
        return this.modelEmpty;
    }

    public set setEmptyMap(modelEmpty: BitwiseMatrix)
    {
        this.modelEmpty = modelEmpty;
    }

    public getValue(y: number, x: number): number
    {
        return getValue(this.data, y, x, this.height, this.width);
    }

    public getIsBoundary(y: number, x: number): boolean
    {
        return getIsBoundary(this.data, y, x, this.height, this.width);
    }

    public countUnrevealedItemsAround(y: number, x: number): number
    {
        return countUnrevealedItemsAround(this.data, y, x, this.height, this.width);
    }

    public getTotalItemsAround(y: number, x: number): number
    {
        return getTotalItemsAround(y, x, this.height, this.width);
    }

    public get clone(): Board
    {
        const board: Board = new Board(this.data, false);
        board.setMinesMap = this.modelMines;
        board.setEmptyMap = this.modelEmpty;

        return board;
    }

    public get isConsistent(): boolean
    {
        for (let i: number = 0; i < this.height; i++) {
            for (let j: number = 0; j < this.width; j++) {
                if (this.getValue(i, j) < 0) {
                    continue;
                }

                if (this.getValue(i, j) < this.modelMines.countAround(i, j)) {
                    return false;
                }

                if (this.getTotalItemsAround(i, j) - this.modelEmpty.countAround(i, j) < this.getValue(i, j)) {
                    return false;
                }
            }
        }

        return true;
    }

    private static isValid(state: number[][]): boolean
    {
        return isValid(state);
    }
}
