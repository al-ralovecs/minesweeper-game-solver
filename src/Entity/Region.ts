import Coordinate from './Coordinate';
import Board from './Board';

export default class Region {
    private data: Coordinate[][] = [];
    private readonly bruteForceLimit: number;

    public constructor(bruteForceLimit: number = 8)
    {
        this.bruteForceLimit = bruteForceLimit;
    }

    public get(n: number): Coordinate[]
    {
        if (0 <= n && this.count > n) {
            return this.data[n];
        }

        throw Error();
    }

    public get count(): number
    {
        return this.data.length;
    }

    public segregate(board: Board): void
    {
        let emptyItems: Coordinate[] = board.listEmptyItems;
        let borderItems: Coordinate[] = board.listBorderItems;

        if (emptyItems.length - borderItems.length <= this.bruteForceLimit) {
            this.data.push(emptyItems);
        } else {
            this.data = Region.process(borderItems);
        }
    }

    private static process(borderItems: Coordinate[]): Coordinate[][]
    {

    }
}
