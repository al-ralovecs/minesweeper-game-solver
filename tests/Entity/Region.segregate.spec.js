import Board from '../../src/Entity/Board';
import Region from '../../src/Entity/Region';
import state from '../fixtures/state/11x16.json';

describe('Region', () => {
    test('check if segregates', () => {
        const board = new Board(state);
        const regions = new Region();
        regions.segregate(board);

        expect(regions.count).toBe(2);
        expect(regions.getData).toMatchSnapshot();
    })
});
