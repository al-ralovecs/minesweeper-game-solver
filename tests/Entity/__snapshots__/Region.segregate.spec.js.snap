// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Region check if segregates 1`] = `
Array [
  Array [
    Coordinate {
      "x": 6,
      "y": 0,
    },
    Coordinate {
      "x": 6,
      "y": 1,
    },
    Coordinate {
      "x": 6,
      "y": 3,
    },
    Coordinate {
      "x": 6,
      "y": 5,
    },
    Coordinate {
      "x": 7,
      "y": 7,
    },
    Coordinate {
      "x": 7,
      "y": 8,
    },
    Coordinate {
      "x": 7,
      "y": 6,
    },
    Coordinate {
      "x": 7,
      "y": 5,
    },
    Coordinate {
      "x": 6,
      "y": 4,
    },
    Coordinate {
      "x": 6,
      "y": 2,
    },
  ],
  Array [
    Coordinate {
      "x": 7,
      "y": 10,
    },
    Coordinate {
      "x": 7,
      "y": 12,
    },
    Coordinate {
      "x": 7,
      "y": 14,
    },
    Coordinate {
      "x": 6,
      "y": 14,
    },
    Coordinate {
      "x": 6,
      "y": 15,
    },
    Coordinate {
      "x": 7,
      "y": 13,
    },
    Coordinate {
      "x": 7,
      "y": 11,
    },
  ],
]
`;
